design/publishing.md

```diff
@@ -0,0 +1,69 @@
+##### This design document focuses on the following - 
+1. Separate entry points for node and browser.
+2. Specifying the browser field in package.json.
+3. Changes in the bundling process.
+
+##### Terms -
+
+* bundler - Module bundlers are tools frontend developers used to bundle JavaScript modules into a single JavaScript files that can be executed in the browser.
+
+* rollup - rollup.js is the module bundler that the JS SDK uses. 
+
+* package.json fields - 
+  * main - The main field is a module ID that is the primary entry point to your program. Points to the CJS modules.
+
+  * module - The module field is not an official npm feature but a common convention among bundlers to designate how to import an ESM version of a library. Points to the ES modules.
+
+  * browser - If the module is meant to be used client-side, the browser field should be used instead of the main field.
+
+##### Current set up -
+
+1. 
+                        TypeScript Source Code
+                                / \
+                      Transpiles into JavaScript 
+                             'lib' folder
+                               /   \
+                     CJS module     ES modules
+2. main - `lib/src/index.js`
+   module - `lib/es/src/index.js`
+
+3. Rollup bundling output
+* `graph-js-sdk.js` - Bundled and minified file in IIFE format. This file can be directly used in the browser with a `<script>` tag.
+* `graph-es-sdk.js` - Bundled filein ES format.
+4. Entry point for rollup - `lib/es/src/browser/index.js`.
+
+##### Difference between src/index.js and src/browser/index.js
+1. `src/browser/index.js` does not export  `RedirectHandler` and `RedirectHandlerOptions`. Redirection is handled by the browser.
+2. `src/browser/index.js` exports `src/browser/ImplicitMsalProvider`.
+3. `src/browser/ImplicitMsalProvider` does not import or require 'msal' dependency. While, 
+`src/ImplicitMsalProvider` imports or requires 'msal' in the implementation. 
+4. My assumtion is that `src/browser/ImplicitMsalProvider` is implemented specifically for the rollup process and to skip the rollup external dependency while using `graph-js-sdk.js` in the browser.
```

Comment:
we shouldn't need both. The MSAL dependency should be identified as an external dependency by the bundler and referenced but not included in the resulting bundle. If this is not the case today, there's probably a configuration with our bundler.

Reply:
Can you point to the configuration in the bundler? 
 
An external dependency should be marked in the `rollup.config`  file which is not done currently. I agree that dependency should be noted in the roll up process going forward to maintain consistency between the modules and the bundled files.

Reply:
I'm not super familiar with rollup but [this](https://stackoverflow.com/a/41432115/3808675) seems to be the way to do it.

Reply:
I assumed you were talking about a config in our bundler set up, something that I might have missed.

Reply:
webpack has an externals configuration but we're not using webpack :)